### ISSUE TITLE
N@tM DVASS Plan

### ISSUE THREAD

**Issue by drewreed2005** [SRIHITAKOTT1213/DVASS#4]
## What Will DVASS Card Games Be?

At the moment, DVASS has frontend-only card game simulations that have been programmed to connect to a backend when possible. By Night at the Museum, we will have the following features up and ready.

### Finishing the Blackjack Game

[Here](https://srihitakott1213.github.io/DVASS/bjsim/) is the current version of blackjack on our site. At the moment, the Blackjack Game has:
- Fully functional frontend gameplay
- Code that allows for connection to a local server database for usernames on the leaderboard
- Passible frontend styling

By the time N@tM comes up, we want to upgrade that to:
- [ ] Full connectivity with a backend leaderboard that saves streak scores
- [ ] More appealing frontend styling, including actual cards instead of text
- [ ] A rules explanation on the page

We plan to integrate that into a more central style that is used throughout the website so that all of the card games look like they're from the same site.

### New Card Games

Since just blackjack isn't a lot of content to offer, we plan on adding three new card games to the site, each with their own unique leaderboard connectivity.

#### Uno

The first game we plan to add to the site is Uno. We already have an in-progress version of the game [here](https://drewreed2005.github.io/realdrew/unohead), and it will likely be fully functional by early next week. Here are the criteria in list form:
- [ ] Fully functional frontend gameplay
- [ ] A system to time the player's match from start to finish for leaderboard entry
- [ ] Full backend leaderboard connectivity
- [ ] Rules explanation
- [ ] Updated frontend styling

#### Go Fish

Since it wouldn't be too hard to use the system we already tested with making interactive, appended HTML elements to make cards for Go Fish, we plan on making it too. This would be another streak-based leaderboard game. The features are listed below:
- [ ] Fully functional frontend gameplay
- [ ] A system to track the player's win streak for the leaderboard
- [ ] Full backend leaderboard connectivity
- [ ] Rules explanation
- [ ] Unique frontend styling (related to fish/fishing)

#### Memory

This is the most unique of the games we'd be making, and would require the most unique code. However, our group has discussed how we will go about making it. In lieu of an entire checklist (since it would basically be the same criteria), this is our current construction plan...
- Create a card class
- Create a number of randomly-generated cards that can be doubled to create a square number (18 most likely)
- Create duplicates of each of the cards
- Randomly select cards to fill each of the rows (6 most likely)
- Flip a certain number of the cards to show the player as a starting point
- POSSIBLY: Only a certain number of times you can get the guess wrong before losing, probably ~5

Because this would be a speed game, we would artificially delay the player's ability to select more cards after selecting incorrectly so that there is a penalty.

### The Leaderboard

[This](https://github.com/drewreed2005/DVASSbackend) is the backend repository we currently have.

The leaderboard will be divided into two categories: streak games and speed games. Blackjack and Go Fish will be streak games because of the nature of those games, while Uno and Memory will be timed, incentivizing players to think fast.

There will be a separate page for the leaderboard. **Looking ahead, this would be where we implement data sorting in our project.** We plan to make the leaderboard have the functions of [this one](https://nighthawkcoders.github.io/APCSP/frontend/jquery) shown in class but with unique styling and potentially other features. This means:
- [ ] Leaderboard ordered by score in backend before the fetch
- [ ] Sort options on the headers themselves
- [ ] A search bar to find specific users
- [ ] (WITH POTENTIAL EXTRA) The ability to find other player scores if the account system is used

### Possible Extras

We have considered two potential extra additions to the project if time allows.

#### Memory Game Difficulty Settings

This would mean Easy, Medium and Hard modes for the memory game. Aside from the additional frontend game code, this would also require multiple leaderboards to be created.

#### User Log-in Functionality

As alluded to before, we've considered potentially creating a system where users have password-protected accounts to save their scores to so that people cannot claim to be someone else on the leaderboard.

The main issue with this is that we'd need to find a way to make the player login to persist in each of the game pages unless we wanted them to log in before each game, which seems like a hassle. Without this system, there isn't much risk of people claiming to be others, however, since there's little incentive to give someone else the credit for your own high score. New low scores wouldn't be added to the leaderboard.

## Group Responsibilities and Schedule

| Name | Scrum Role | Responsibilities |
| --- | --- | --- |
| Drew | Scrum Master | Finishing the Uno functionality, creating starters for the other two games, revising and finishing backend model/API, data sorting for leaderboard backend |
| Sreeja | Project Manager | Writing and revising backend model/API files, connecting to frontend (when possible), frontend game code early on |
| Vivian | Frontend Developer | Focus on Memory game code and frontend styling for Uno and Memory games, collaborate with Aliya on leaderboard styling when possible |
| Aliya | Backend Developer | Focus on Go Fish game code and frontend styling for Blackjack and Go Fish games, collaborate with Vivian on leaderboard styling when possible |
| Sri | DevOps | Deployment of the leaderboard backend, Memory game code, frontend game styling |

### (Short) Development Timeline

Week 1: Creating remaining model/API files starts Monday, Uno game code finished by Tuesday, and functional code for Go Fish and Memory by Saturday, begin deployment once backend is done (as early as possible)

Week 2: Get through deployment as soon as possible and connect games, create and style leaderboard, and make finishing touches on game frontend styling

### Group Scrum Board

[Click here](https://github.com/users/drewreed2005/projects/3/views/1) to access the group Scrum Board for this project. All of the things described above are found on the board, plus the extras in their own category.

#### Tracking Collaborations & Time

Each day, group members will add the amount of time they spent working toward a specific goal on the Scrum Board descriptions, along with the date. This will help us give credit where credit is due.
